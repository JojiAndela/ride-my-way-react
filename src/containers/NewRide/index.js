import React, { Component } from 'react';
import { connect } from 'react-redux';

import Form from 'Components/Forms/Form';
import { createRide } from 'Store/actions/rides';
import { updateObject, checkValidity } from '../../utility';

/**
 * NewRide
 */

        //
        // name: name.value,
        // location: location.value,
        // destination: destination.value,
        // time: time.value,
        // takeOffDate: date.value,
        // carModel: car.value,
        // slot: Number(slot.value)
        //
export class NewRide extends Component {

    state={
      rideForm: {
        name: {
            inputType: 'input',
            elementConfig: {
                type: 'text',
                placeholder: 'A name for your ride',
                minLength: "3",
                required: true
            },
            value: '',
            validation: {
                required: true,
            },
            valid: false,
        },
        location: {
          inputType: 'input',
          elementConfig: {
              type: 'text',
              placeholder: 'Where are you going from ?',
              minLength: "3",
              required: true
          },
          value: '',
          validation: {
              required: true
          },
          valid: false,
      },
      destination: {
        inputType: 'input',
        elementConfig: {
            type: 'text',
            placeholder: 'Where are you going to ?',
            minLength: "3",
            required: true
        },
        value: '',
        validation: {
            required: true
        },
        valid: false,
    },
    time: {
      inputType: 'input',
      elementConfig: {
          type: "time",
          min: "05:00",
          max: "21:00",
          placeholder: "enter the departure time ...",
          required: true
      },
      value: '',
      validation: {
          required: true
      },
      valid: false,
    },
    date: {
      inputType: 'input',
      elementConfig: {
          type:"date",
          min:"2018-07-01",
          max:"2020-01-01",
          placeholder:"enter the departure date...",
          required: true
      },
      value: '',
      validation: {
          required: true
      },
      valid: false,
  },
  car: {
    inputType: 'input',
    elementConfig: {
        type: 'text',
        placeholder: 'enter the car Model(vehicle type) ...',
        minLength: "3",
        required: true
    },
    value: '',
    validation: {
        required: true
    },
    valid: false,
},
slot: {
  inputType: 'input',
  elementConfig: {
      type: 'number',
      step: "1",
      min: "1",
      placeholder: "Available seats ...",
      required: true
  },
  value: '',
  validation: {
      required: true
  },
  valid: false,
},
    }
    }
    onBtnClick = (e) => {
      e.preventDefault();
        const rideData = {
          name: this.state.rideForm.name.value,
          location: this.state.rideForm.location.value,
          destination: this.state.rideForm.destination.value,
          time: this.state.rideForm.time.value,
          takeOffDate: this.state.rideForm.date.value,
          carModel: this.state.rideForm.car.value,
          slot: Number(this.state.rideForm.slot.value)
        };
        console.log(rideData);
        return this.props.createRideAction(rideData, this.props.history);
    }

    inputChangedHandler = (event, inputIdentifier) => {

            const updatedFormElement = updateObject(this.state.rideForm[inputIdentifier], {
                value: event.target.value,
                valid: checkValidity(event.target.value, this.state.rideForm[inputIdentifier].validation),
            });
            const updatedOrderForm = updateObject(this.state.rideForm, {
                [inputIdentifier]: updatedFormElement
            });

            let formIsValid = true;
            for (let inputIdentifier in updatedOrderForm) {
                formIsValid = updatedOrderForm[inputIdentifier].valid && formIsValid;
            }
            this.setState({rideForm: updatedOrderForm, formIsValid: formIsValid});
        }
    render() {
      return (
        <div>
            <h1 style={{textAlign: 'center', color: '#bbb'}}>Create A Ride</h1>
          <Form form={this.state.rideForm} btnName="Create Ride" btnClick={this.onBtnClick} onChanged={this.inputChangedHandler}/>
        </div>
      );
  }
  }

  const mapStateToProps = state => ({
    loading: state.rides.loading
  });

  const mapDispatchToProps = dispatch => ({
    createRideAction: (data, history) => dispatch(createRide(data, history)),
  });

  export default connect(mapStateToProps, mapDispatchToProps)(NewRide);
