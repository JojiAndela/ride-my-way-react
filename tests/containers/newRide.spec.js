import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

// Component to be tested
import NewRide from '../../src/containers/NewRide';

const initialState ={
  rides: {
    loading: false,
    rides: [],
    singleRide: null,
    requests: [],
    myRequest: {}
  }
}
const state = {rideForm:{
  name: {
      inputType: 'input',
      elementConfig:{
          type: 'text',
          placeholder: 'Your Username',
          minLength: "3",
          required: true,
      },
      value: '',
      validation: {
          required: true
      },
      valid: false,
      touched: false,
  },
}}

const onBtnClick = jest.fn();
const inputChangedHandler = jest.fn();
const middleware = [thunk];
const mockStore = configureStore(middleware);
const createRideAction = jest.fn();
const setErrorAction = jest.fn();

let wrapper;
let store;
describe('NewRide', () => {
  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <BrowserRouter>
          <NewRide onBtnClick={onBtnClick}
            inputChangedHandler={inputChangedHandler}
            createRideAction={createRideAction}
            setErrorAction={setErrorAction}
          />
        </BrowserRouter>
      </Provider>
    );
  });
  it('should render the new ride page correctly', () => {
    wrapper.setState(state);
    expect(wrapper).toMatchSnapshot();
  });
});
