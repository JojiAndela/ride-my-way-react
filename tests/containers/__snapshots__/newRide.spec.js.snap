// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewRide should render the new ride page correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(NewRide)
        createRideAction={[MockFunction]}
        inputChangedHandler={[MockFunction]}
        onBtnClick={[MockFunction]}
        setErrorAction={[MockFunction]}
      >
        <NewRide
          createRideAction={[Function]}
          inputChangedHandler={[MockFunction]}
          loading={false}
          onBtnClick={[MockFunction]}
          setErrorAction={[MockFunction]}
        >
          <div>
            <h1
              style={
                Object {
                  "color": "#bbb",
                  "textAlign": "center",
                }
              }
            >
              Create A Ride
            </h1>
            <form
              btnClick={[Function]}
              btnName="Create Ride"
              form={
                Object {
                  "car": Object {
                    "elementConfig": Object {
                      "minLength": "3",
                      "placeholder": "enter the car Model(vehicle type) ...",
                      "required": true,
                      "type": "text",
                    },
                    "inputType": "input",
                    "valid": false,
                    "validation": Object {
                      "required": true,
                    },
                    "value": "",
                  },
                  "date": Object {
                    "elementConfig": Object {
                      "max": "2020-01-01",
                      "min": "2018-07-01",
                      "placeholder": "enter the departure date...",
                      "required": true,
                      "type": "date",
                    },
                    "inputType": "input",
                    "valid": false,
                    "validation": Object {
                      "required": true,
                    },
                    "value": "",
                  },
                  "destination": Object {
                    "elementConfig": Object {
                      "minLength": "3",
                      "placeholder": "Where are you going to ?",
                      "required": true,
                      "type": "text",
                    },
                    "inputType": "input",
                    "valid": false,
                    "validation": Object {
                      "required": true,
                    },
                    "value": "",
                  },
                  "location": Object {
                    "elementConfig": Object {
                      "minLength": "3",
                      "placeholder": "Where are you going from ?",
                      "required": true,
                      "type": "text",
                    },
                    "inputType": "input",
                    "valid": false,
                    "validation": Object {
                      "required": true,
                    },
                    "value": "",
                  },
                  "name": Object {
                    "elementConfig": Object {
                      "minLength": "3",
                      "placeholder": "A name for your ride",
                      "required": true,
                      "type": "text",
                    },
                    "inputType": "input",
                    "valid": false,
                    "validation": Object {
                      "required": true,
                    },
                    "value": "",
                  },
                  "slot": Object {
                    "elementConfig": Object {
                      "min": "1",
                      "placeholder": "Available seats ...",
                      "required": true,
                      "step": "1",
                      "type": "number",
                    },
                    "inputType": "input",
                    "valid": false,
                    "validation": Object {
                      "required": true,
                    },
                    "value": "",
                  },
                  "time": Object {
                    "elementConfig": Object {
                      "max": "21:00",
                      "min": "05:00",
                      "placeholder": "enter the departure time ...",
                      "required": true,
                      "type": "time",
                    },
                    "inputType": "input",
                    "valid": false,
                    "validation": Object {
                      "required": true,
                    },
                    "value": "",
                  },
                }
              }
              onChanged={[Function]}
            >
              <form
                className="Form"
                onSubmit={[Function]}
              >
                <input
                  changed={[Function]}
                  elementConfig={
                    Object {
                      "minLength": "3",
                      "placeholder": "A name for your ride",
                      "required": true,
                      "type": "text",
                    }
                  }
                  inputType="input"
                  key="name"
                  value=""
                >
                  <input
                    className="input"
                    minLength="3"
                    onChange={[Function]}
                    placeholder="A name for your ride"
                    required={true}
                    type="text"
                    value=""
                  />
                </input>
                <input
                  changed={[Function]}
                  elementConfig={
                    Object {
                      "minLength": "3",
                      "placeholder": "Where are you going from ?",
                      "required": true,
                      "type": "text",
                    }
                  }
                  inputType="input"
                  key="location"
                  value=""
                >
                  <input
                    className="input"
                    minLength="3"
                    onChange={[Function]}
                    placeholder="Where are you going from ?"
                    required={true}
                    type="text"
                    value=""
                  />
                </input>
                <input
                  changed={[Function]}
                  elementConfig={
                    Object {
                      "minLength": "3",
                      "placeholder": "Where are you going to ?",
                      "required": true,
                      "type": "text",
                    }
                  }
                  inputType="input"
                  key="destination"
                  value=""
                >
                  <input
                    className="input"
                    minLength="3"
                    onChange={[Function]}
                    placeholder="Where are you going to ?"
                    required={true}
                    type="text"
                    value=""
                  />
                </input>
                <input
                  changed={[Function]}
                  elementConfig={
                    Object {
                      "max": "21:00",
                      "min": "05:00",
                      "placeholder": "enter the departure time ...",
                      "required": true,
                      "type": "time",
                    }
                  }
                  inputType="input"
                  key="time"
                  value=""
                >
                  <input
                    className="input"
                    max="21:00"
                    min="05:00"
                    onChange={[Function]}
                    placeholder="enter the departure time ..."
                    required={true}
                    type="time"
                    value=""
                  />
                </input>
                <input
                  changed={[Function]}
                  elementConfig={
                    Object {
                      "max": "2020-01-01",
                      "min": "2018-07-01",
                      "placeholder": "enter the departure date...",
                      "required": true,
                      "type": "date",
                    }
                  }
                  inputType="input"
                  key="date"
                  value=""
                >
                  <input
                    className="input"
                    max="2020-01-01"
                    min="2018-07-01"
                    onChange={[Function]}
                    placeholder="enter the departure date..."
                    required={true}
                    type="date"
                    value=""
                  />
                </input>
                <input
                  changed={[Function]}
                  elementConfig={
                    Object {
                      "minLength": "3",
                      "placeholder": "enter the car Model(vehicle type) ...",
                      "required": true,
                      "type": "text",
                    }
                  }
                  inputType="input"
                  key="car"
                  value=""
                >
                  <input
                    className="input"
                    minLength="3"
                    onChange={[Function]}
                    placeholder="enter the car Model(vehicle type) ..."
                    required={true}
                    type="text"
                    value=""
                  />
                </input>
                <input
                  changed={[Function]}
                  elementConfig={
                    Object {
                      "min": "1",
                      "placeholder": "Available seats ...",
                      "required": true,
                      "step": "1",
                      "type": "number",
                    }
                  }
                  inputType="input"
                  key="slot"
                  value=""
                >
                  <input
                    className="input"
                    min="1"
                    onChange={[Function]}
                    placeholder="Available seats ..."
                    required={true}
                    step="1"
                    type="number"
                    value=""
                  />
                </input>
                <button
                  btnType="Success"
                >
                  <button
                    className="Button Success"
                  >
                    Create Ride
                  </button>
                </button>
              </form>
            </form>
          </div>
        </NewRide>
      </Connect(NewRide)>
    </Router>
  </BrowserRouter>
</Provider>
`;
